src/lib/messaging.ts

```diff
@@ -79,23 +83,80 @@ function backgroundHandler<
+    //                                     Type 'ReturnType<Root[string][string]>' is not assignable to type 'Promise<any>'.
+    //                                       Type 'ReturnType<Root[string | number | symbol][string]>' is not assignable to type 'Promise<any>'.
+    //                                         Type 'ReturnType<Root[keyof Root][string]>' is not assignable to type 'Promise<any>'.
+    //                                           Type 'ReturnType<Root[keyof Root][string | number | symbol]>' is not assignable to type 'Promise<any>'.
+    //                                             Type 'ReturnType<Root[keyof Root][keyof Root[keyof Root]]>' is not assignable to type 'Promise<any>'.
+    //                                               Type 'unknown' is not assignable to type 'Promise<any>'.
+    //                                                 Type 'ReturnType<Root[keyof Root][string | number | symbol]>' is not assignable to type 'Promise<any>'.
+    //                                                   Type 'unknown' is not assignable to type 'Promise<any>'.
+    //                                                     Type 'ReturnType<Root[keyof Root][string]> | ReturnType<Root[keyof Root][number]> | ReturnType<Root[keyof Root][symbol]>' is not assignable to type 'Promise<any>'.
+    //                                                       Type 'ReturnType<Root[keyof Root][string]>' is not assignable to type 'Promise<any>'.
+    //                                                         Type 'unknown' is not assignable to type 'Promise<any>'.
+    //                                                           Type 'ReturnType<Root[string | number | symbol][string]>' is not assignable to type 'Promise<any>'.
+    //                                                             Type 'unknown' is not assignable to type 'Promise<any>'.
+    //                                                               Type 'ReturnType<Root[string][string]> | ReturnType<Root[number][string]> | ReturnType<Root[symbol][string]>' is not assignable to type 'Promise<any>'.
+    //                                                                 Type 'ReturnType<Root[string][string]>' is not assignable to type 'Promise<any>'.
+    //                                                                   Type '{}' is missing the following properties from type 'Promise<any>': then, catch, [Symbol.toStringTag], finally
+    //
+    // ERROR in /home/olie/projects/tridactyl/src/lib/messaging.ts
+    // ./src/lib/messaging.ts
+    // [tsl] ERROR in /home/olie/projects/tridactyl/src/lib/messaging.ts(115,40)
+    //       TS2558: Expected 0 type arguments, but got 2.
+    // ```
+    //
+    // don't you understand?
+    //
+    // (This is why there is an `: any => `)
+    browser.runtime.onMessage.addListener(
+        (message: any, sender: browser.runtime.MessageSender): any => {
```

Comment:
@farnoy I think this function is yours. Do you know why DefinitelyTyped hates it so much?

Reply:
@bovine3dom Sorry about the delay. The callback to `addListener` should return a boolean or a Promise, but MDN docs advise against using `async` closure here directly. I think we need to modify `backgroundHandler`, or do something like `return !!backgroundHandler(root, message, sender)` to cast this to a boolean

Reply:
Actually `return Promise.resolve(backgroundHandler(root, message, sender))` should be enough, but I'm not sure if it's going to typecheck

Reply:
That seems to work, thanks - 27c5a036b424020d49b9a1bf13206ca753471b4d